### PR TITLE
Coverage Metrics

### PR DESCRIPTION
**Pull request by ztnel**
# Description
Adding coverage metrics to `simpledbus`. I would also like to include integration with codecov.io to compliment coverage metrics.

To build `simpledbus_test` with coverage set the `CMAKE_COVERAGE` flag:
```bash
cmake -DCMAKE_COVERAGE=On -DCMAKE_BUILD_TYPE=Debug ../test
```

Then run the `gcov` target to run the tests and generate a coverage report:
```bash
make gcov
```

The `gcov` target will generate a cli report (for local unittesting) and an xml file for ci builds and pushes to a third party code coverage platform.

## Checklist
 - [x] Added code coverage cmake build mode
 - [x] Update README to include coverage build mode
 - [x] Update build test scripts to include coverage build flag
 - [x] Add build and push action tasks for coverage metrics

## Notes
Thread Sanitizer unittests fail with coverage compiler flags set:
```
SUMMARY: ThreadSanitizer: data race /usr/include/c++/9/bits/atomic_base.h:100 in std::operator&(std::memory_order, std::__memory_order_modifier)
[38](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:38)
==================
[39](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:39)
==================
[40](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:40)
WARNING: ThreadSanitizer: data race (pid=4720)
[41](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:41)
  Read of size 8 at 0x564910218128 by thread T1:
[42](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:42)
    #0 std::operator&(std::memory_order, std::__memory_order_modifier) /usr/include/c++/9/bits/atomic_base.h:101 (simpledbus_test+0x14e73c)
[43](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:43)
    #1 std::atomic<bool>::operator=(bool) /usr/include/c++/9/atomic:81 (simpledbus_test+0x14e921)
[44](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:44)
    #2 PythonRunner::_async_thread_function() /home/runner/work/SimpleDBus/SimpleDBus/test/src/helpers/PythonRunner.cpp:[45](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:45) (simpledbus_test+0x14df0c)
45
    #3 void std::__invoke_impl<void, void (PythonRunner::*)(), PythonRunner*>(std::__invoke_memfun_deref, void (PythonRunner::*&&)(), PythonRunner*&&) /usr/include/c++/9/bits/invoke.h:73 (simpledbus_test+0x158345)
[46](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:46)
    #4 std::__invoke_result<void (PythonRunner::*)(), PythonRunner*>::type std::__invoke<void (PythonRunner::*)(), PythonRunner*>(void (PythonRunner::*&&)(), PythonRunner*&&) /usr/include/c++/9/bits/invoke.h:95 (simpledbus_test+0x157ed1)
[47](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:47)
    #5 void std::thread::_Invoker<std::tuple<void (PythonRunner::*)(), PythonRunner*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) /usr/include/c++/9/thread:244 (simpledbus_test+0x157ac5)
[48](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:48)
    #6 std::thread::_Invoker<std::tuple<void (PythonRunner::*)(), PythonRunner*> >::operator()() /usr/include/c++/9/thread:251 (simpledbus_test+0x1578c2)
[49](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:49)
    #7 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (PythonRunner::*)(), PythonRunner*> > >::_M_run() /usr/include/c++/9/thread:195 (simpledbus_test+0x15779e)
[50](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:50)
    #8 <null> <null> (libstdc++.so.6+0xd6de3)
[51](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:51)

[52](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:52)
  Previous write of size 8 at 0x564910218128 by main thread:
[53](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:53)
    #0 std::operator&(std::memory_order, std::__memory_order_modifier) /usr/include/c++/9/bits/atomic_base.h:101 (simpledbus_test+0x14e753)
[54](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:54)
    #1 PythonRunner::init() /home/runner/work/SimpleDBus/SimpleDBus/test/src/helpers/PythonRunner.cpp:17 (simpledbus_test+0x14d421)
[55](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:55)
    #2 main /home/runner/work/SimpleDBus/SimpleDBus/test/src/main.cpp:9 (simpledbus_test+0x5af6a)
[56](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:56)

[57](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:57)
  Location is global '__gcov0._ZStanSt12memory_orderSt23__memory_order_modifier' of size 16 at 0x564910218120 (simpledbus_test+0x000000318128)
[58](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:58)

[59](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:59)
  Thread T1 (tid=4722, running) created by main thread at:
[60](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:60)
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:962 (libtsan.so.0+0x5ea79)
[61](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:61)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd70a8)
[62](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:62)
    #2 PythonRunner::init() /home/runner/work/SimpleDBus/SimpleDBus/test/src/helpers/PythonRunner.cpp:16 (simpledbus_test+0x14d3d3)
[63](https://github.com/ztnel/SimpleDBus/runs/5267313759?check_suite_focus=true#step:13:63)
    #3 main /home/runner/work/SimpleDBus/SimpleDBus/test/src/main.cpp:9 (simpledbus_test+0x5af6a)
``` 

# Dependencies
- `gcovr>=5.0`

### CODE REVIEW

File: scripts/build_tests.sh [hunk 1]
```diff
@@ -21,5 +21,5 @@ fi
 # Compile!
 mkdir -p $BUILD_PATH
 cd $BUILD_PATH
-cmake -H$PROJECT_ROOT/test -DCMAKE_BUILD_TYPE=Debug
+cmake -H$PROJECT_ROOT/test -DCMAKE_BUILD_TYPE=Debug -DCMAKE_COVERAGE=On
```
Comment (kdewald) [inline scripts/build_tests.sh:24]: `CMAKE_COVERAGE` is not an official CMake option. You should never do this, it is confusing.
You can use `SIMPLEDBUS_COVERAGE` instead.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,11 @@ project(simpledbus_test)
 
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_OUTPUT_EXTENSION_REPLACE ON)
+
+if(CMAKE_COVERAGE MATCHES "On")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -O0 -fprofile-arcs -ftest-coverage")
```
Comment (kdewald) [inline test/CMakeLists.txt:9]: Don't these flags directly, that's bad practice. They affect all the downstream projects that consume this library. See how it was done with the other targets.

Also, when done like this, these flags will never get propagated to the upstream libraries.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -32,3 +37,25 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
...
   COMMAND "${CMAKE_COMMAND}" -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/python/ ${CMAKE_BINARY_DIR}
 )
+
+if(CMAKE_COVERAGE MATCHES "On")
+  find_program(GCOVR_PATH gcovr)
+  if(NOT GCOVR_PATH)
+    message(FATAL_ERROR "gcovr not found! Aborting...")
+  endif() # NOT GCOVR_PATH
+  # Create the gcov target. Run coverage tests with 'make gcov'
+  add_custom_target(gcov
```
Comment (kdewald) [inline test/CMakeLists.txt:47]: Never use target names that might overlap with system programs or common names. These targets are visible globally, so they should be prefixed. `simpledbus_test_coverage` is appropriate.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -32,3 +37,25 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
...
+  add_custom_command(TARGET gcov
+    # cli report for local testing
+    COMMAND echo "Generating cli coverage report"
+    COMMAND gcovr -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/..
+    # xml report for ci
+    COMMAND echo "Generating xml coverage report"
+    COMMAND gcovr --xml -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/.. -o cov.xml
+  )
+  add_dependencies(gcov simpledbus_test)
+  add_test(output_test ${CMAKE_BINARY_DIR}/simpledbus_test)
```
Comment (kdewald) [inline test/CMakeLists.txt:60]: Why is this necessary?

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -32,3 +37,25 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
...
+endif() # CMAKE_BUILD_TYPE MATCHES "Coverage"
```
Comment (kdewald) [inline test/CMakeLists.txt:61]: Always newline at the end of files.

File: README.md [hunk 6]
```diff
@@ -47,6 +47,16 @@ make -j
...
+To build with test coverage set the `CMAKE_COVERAGE` flag:
+```bash
+cmake -DCMAKE_COVERAGE=On -DCMAKE_BUILD_TYPE=Debug ../test
+```
+
+Then run the `gcov` target to run the tests and generate an inline coverage report:
+```bash
+make gcov
+```
+
```
Comment (kdewald) [inline README.md:59]: This would have to go on its own section.

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -32,3 +37,25 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
 add_custom_command (TARGET simpledbus_test POST_BUILD
   COMMAND "${CMAKE_COMMAND}" -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/python/ ${CMAKE_BINARY_DIR}
 )
+
+if(CMAKE_COVERAGE MATCHES "On")
+  find_program(GCOVR_PATH gcovr)
+  if(NOT GCOVR_PATH)
+    message(FATAL_ERROR "gcovr not found! Aborting...")
+  endif() # NOT GCOVR_PATH
```
Comment (kdewald) [inline test/CMakeLists.txt:45]: I'm not a fan of commenting the contents of the `if` statement here.

**Comment by kdewald** [comment]
On the error you see when running thread sanitizer with coverage, where do you think it might be coming from?

File: test/CMakeLists.txt (hunk 2 above)
Comment (ztnel) [reply test/CMakeLists.txt:9]: good to know. I assume you are referring to `target_compile_options`

File: test/CMakeLists.txt (hunk 4 above)
Comment (ztnel) [reply test/CMakeLists.txt:60]: its not. I can just run against the binary directly.

**Comment by ztnel** [comment]
> 

To be honest the thread sanitizer test goes beyond my understanding. Maybe a brief overview would help me understand the problem better.

**Comment by ztnel** [comment]
> On the error you see when running thread sanitizer with coverage, where do you think it might be coming from?

To be honest the thread sanitizer test goes beyond my understanding. Maybe a brief overview would help me understand the problem better.

File: README.md (hunk 6 above)
Comment (ztnel) [reply README.md:59]: Yea makes sense. I also think the readme in the root is too large could maybe be divided into multiple readmes linked together by nav hyperlinks kinda like so: https://github.com/OpenHerb/telemeter/blob/master/docs/README.md

what are your thoughts?

Comment (kdewald) [reply README.md]: I agree that we should organize documentation a bit better.
I already set these up https://readthedocs.org/projects/simpledbus/ for all the projects in the organization. Now it's just a matter of writing all the documentation xD

(I'm doing a bit of additional research to determine the best tooling for this, but I'll start working on this soon.)

File: CMakeLists.txt [hunk 8]
```diff
@@ -40,7 +40,7 @@ if(NOT LIBFMT_LOCAL_PATH)
     )
     FetchContent_MakeAvailable(fmt)
 else()
-    message("-- [INFO] Using local LibFmt")
+    message(STATUS "[INFO] Using local LibFmt")
```
Comment (kdewald) [inline CMakeLists.txt:43]: Let's try to keep each pull request as small as possible. I agree with this change, but I'd rather we ship this on a different MR, so as to keep conversations on topic.

File: README.md [hunk 9]
```diff
@@ -47,6 +47,17 @@ make -j
...
 
+### Coverage
+To build with test coverage set the `SIMPLEDBUS_COVERAGE` flag:
+```bash
+cmake -DSIMPLEDBUS_COVERAGE=On -DCMAKE_BUILD_TYPE=Debug ../test
+```
+
+Then run the `gcov` target to run the tests and generate an inline coverage report:
+```bash
+make gcov
```
Comment (kdewald) [inline README.md:58]: Reminder to change the name of this target.

File: .github/workflows/ci_test.yml [hunk 10]
```diff
@@ -30,17 +30,20 @@ jobs:
...
-          cmake -DCMAKE_BUILD_TYPE=Release ../test
+          cmake -DCMAKE_BUILD_TYPE=Release -DSIMPLEDBUS_COVERAGE=On ../test
           make -j4
-      - name: Run Unit Tests
-        run: |
-          cd $GITHUB_WORKSPACE/build
-          ./simpledbus_test
+      - name: Run Unit Tests with Coverage
+        working-directory: build
+        run: make gcov
```
Comment (kdewald) [inline .github/workflows/ci_test.yml:42]: Reminder to change the name of the target.

Also add the `-j4` parameter to make the build faster.

File: test/CMakeLists.txt [hunk 11]
```diff
@@ -3,6 +3,7 @@ project(simpledbus_test)
 
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_OUTPUT_EXTENSION_REPLACE ON)
```
Comment (kdewald) [inline test/CMakeLists.txt:6]: Can you add a quick comment explaining why this is necessary?

File: test/CMakeLists.txt [hunk 12]
```diff
@@ -26,9 +27,31 @@ set(
 )
 
 add_executable(simpledbus_test ${SRC_FILES})
+
```
Comment (kdewald) [inline test/CMakeLists.txt:30]: nit: Extra newline

File: test/CMakeLists.txt [hunk 13]
```diff
@@ -26,9 +27,31 @@ set(
...
 target_link_libraries(simpledbus_test simpledbus-static ${GTEST_LIBRARIES} ${Python3_LIBRARIES} pthread)
 target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Python3_INCLUDE_DIRS})
 
 add_custom_command (TARGET simpledbus_test POST_BUILD
   COMMAND "${CMAKE_COMMAND}" -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/python/ ${CMAKE_BINARY_DIR}
 )
+
+if(SIMPLEDBUS_COVERAGE MATCHES "On")
+  set(SIMPLEDBUS_GCOV_TARGET ${PROJECT_NAME}_coverage)
```
Comment (kdewald) [inline test/CMakeLists.txt:39]: There's no need to create this variable. You can just use `simpledbus_coverage` wherever needed.

File: test/CMakeLists.txt [hunk 14]
```diff
@@ -26,9 +27,31 @@ set(
...
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+  )
+  add_custom_command(TARGET ${SIMPLEDBUS_GCOV_TARGET}
+    # cli report for local testing
+    COMMAND echo "Generating cli coverage report"
+    COMMAND gcovr -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/..
+    # xml report for ci
+    COMMAND echo "Generating xml coverage report"
+    COMMAND gcovr --xml -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/.. -o cov.xml
+  )
```
Comment (kdewald) [inline test/CMakeLists.txt:55]: Could these two statements be merged into one?

File: test/CMakeLists.txt [hunk 15]
```diff
@@ -26,9 +27,31 @@ set(
...
+  find_program(GCOVR_PATH gcovr)
+  if(NOT GCOVR_PATH)
+    message(FATAL_ERROR "gcovr not found! Aborting...")
+  endif()
+  add_custom_target(${SIMPLEDBUS_GCOV_TARGET}
+    COMMAND GTEST_OUTPUT=xml:results/ GTEST_COLOR=1 ./simpledbus_test
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+  )
+  add_custom_command(TARGET ${SIMPLEDBUS_GCOV_TARGET}
+    # cli report for local testing
```
Comment (kdewald) [inline test/CMakeLists.txt:49]: nit: `CLI` in uppercase

File: test/CMakeLists.txt [hunk 16]
```diff
@@ -26,9 +27,31 @@ set(
...
+    COMMAND echo "Generating cli coverage report"
+    COMMAND gcovr -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/..
+    # xml report for ci
```
Comment (kdewald) [inline test/CMakeLists.txt:52]: nit: `XML` and `CI` in uppercase

**Comment by kdewald** [comment]
@ztnel I've also enabled the CI pipeline on this MR, so you can now ensure that the tests pass before landing.

File: CMakeLists.txt (hunk 8 above)
Comment (ztnel) [reply CMakeLists.txt:43]: mm gotcha a bad habit of mine

File: test/CMakeLists.txt (hunk 14 above)
Comment (ztnel) [reply test/CMakeLists.txt:55]: Yes can just create a custom target.

File: test/CMakeLists.txt (hunk 11 above)
Comment (ztnel) [reply test/CMakeLists.txt:6]: mm yea basically gcovs discovery of `.gcda` and `.gcno` files (coverage artefacts) are impeded by the `.cpp` file extensions. By this extent would it be better to hid this configuration under the `SIMPLEDBUS_COVERAGE` flag?

Comment (kdewald) [reply test/CMakeLists.txt:7]: With the comment it's understandable enough, no need to move this declaration.

File: test/CMakeLists.txt [hunk 17]
```diff
@@ -32,3 +34,20 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
...
+
+if(SIMPLEDBUS_COVERAGE MATCHES "On")
+  find_program(GCOVR_PATH gcovr)
+  if(NOT GCOVR_PATH)
+    message(FATAL_ERROR "gcovr not found! Aborting...")
+  endif()
+  add_custom_target(${PROJECT_NAME}_coverage
+    COMMAND echo "Executing test suite"
+    COMMAND GTEST_OUTPUT=xml:results/ GTEST_COLOR=1 ./simpledbus_test
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
```
Comment (kdewald) [inline test/CMakeLists.txt:46]: nit: `WORKING_DIRECTORY` is placed at the end

File: README.rst [hunk 18]
```diff
@@ -66,6 +66,22 @@ Build tests
 To build and run unit and integration tests, the following packages are
 required: ``sudo apt install libgtest-dev libgmock-dev python3-dev``
 
+Build Coverage
+~~~~~~~~~~~~~~
```
Comment (kdewald) [inline README.rst:70]: Should be this kind of line: `^^^^^^^^^^^^^^^`

File: README.rst [hunk 19]
```diff
@@ -66,6 +66,22 @@ Build tests
 To build and run unit and integration tests, the following packages are
 required: ``sudo apt install libgtest-dev libgmock-dev python3-dev``
 
+Build Coverage
```
Comment (kdewald) [inline README.rst:69]: Just `Coverage`

File: README.rst [hunk 20]
```diff
@@ -66,6 +66,22 @@ Build tests
...
+To build with test coverage set the ``SIMPLEDBUS_COVERAGE`` flag:
+::
+   
+   cmake -DSIMPLEDBUS_COVERAGE=On -DCMAKE_BUILD_TYPE=Debug ../test
+
+
+Then run the ``simpledbus_test_coverage`` target to run the tests and generate an inline coverage report:
+::
+
+   make simpledbus_test_coverage -j
```
Comment (kdewald) [inline README.rst:81]: In order to run tests with coverage analysis, CMake needs to be called with the following option: `-DSIMPLEDBUS_COVERAGE=On`. The coverage analysis target is called `simpledbus_test_coverage`.

File: test/CMakeLists.txt [hunk 21]
```diff
@@ -32,3 +34,20 @@ target_include_directories(simpledbus_test PRIVATE ${GTEST_INCLUDE_DIRS} ${Pytho
...
+    COMMAND echo "Generating CLI coverage report"
+    COMMAND gcovr -r .. . -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/..
```
Comment (kdewald) [inline test/CMakeLists.txt:48]: Let's be explicit with the options:
`-r` -> `--root`
`-e` -> `--exclude`

Comment (kdewald) [reply test/CMakeLists.txt:48]: Also, you didn't read through the documentation of the options correctly. The command should be:
`gcovr -r ${CMAKE_CURRENT_LIST_DIR}/.. -e ${CMAKE_BINARY_DIR}/_deps -e ${CMAKE_CURRENT_LIST_DIR}/src`

I'll list all the changes:
- I removed the `.`. It was redundant. If it weren't, it would have been at the end.
- Your root directory for gcovr was set relative to the binary directory, which could be anywhere if this project is being built as part of another. You should make it relative to the current list directory.
- You were excluding the entire root folder too, when instead it should be only `test/src`, so I updated the paths accordingly.

Comment (ztnel) [reply test/CMakeLists.txt]: Nice catch, didn't know the binary dir could change. I updated the commands to reflect these changes

**Comment by kdewald** [comment]
Pull request is looking good. If the tests pass it should be good to land.
